# Used Car Listing Price Prediction

---

<center><img src="https://raw.githubusercontent.com/anthonynamnam/anthonynamnam/main/icons/image/car-banner.png" alt="memes" width="600" /></center>

---

## Project Overview <a class="anchor" id="overview"></a>

Please kindly refer to the github repo of this project: [Link](https://github.com/anthonynamnam/brainstation_capstone#project-overview)


---

## Notebook Overview <a class="anchor" id="overview"></a>

In this notebook, we will go through part of our project, which is the sprint 1 of our project.


---

## Table of Contents <a class="anchor" id="3-toc"></a>

- [Notebook Set Up](#3-setup)
- [Functions](#3-func)
- [Data Loading](#3-load)
- [Feature Transformation](#3-feat-transform)
- [Categorical Encoding](#3-cat-encode)
- [Feature Scaling](#3-feat-scale)
- [Dimensionality Reduction](#3-)
    - PCA
    - MDS
    - tSNE
    - Kernal PCA
- [Feature Selection](#3-)
- [Learning/Takeaway](#3-learn)

---

## Set Up <a class="anchor" id="3-setup"></a>

### Install Library

### Import Library

In [53]:
# Data Science Package
import numpy as np
import pandas as pd



### Global Constant

In [54]:

pd.options.display.max_columns = None

## Functions <a class="anchor" id="3-func"></a>

### Helper Funcntions (Print Info)

In [56]:
# Helper Functions to print df info and statement
import pandas as pd

def print_num_row(df: pd.DataFrame) -> None:
    """
    Description
    -----
    Retrieve the number of rows of dataframe and print it as a statement.
    
    Args
    -----
    df (pd.DataFrame): the target dataframe
    
    Returns
    -----
    None
    
    Example
    -----
    df = pd.DataFrame(data = {"height":[147,190],"weight":[47,72],"age":[12,28]},index = [0,1])
    print_num_row(df)  =>
        |
        | "The dataframe has 2 rows of record now."
        |
    
    
    """
    print(f"The dataframe has {df.shape[0]} rows of record now.")
    return
    

def print_num_col(df: pd.DataFrame) -> None:
    """
    Description
    -----
    Retrieve the number of columns of dataframe and print it as a statement.
    
    Args
    -----
    df (pd.DataFrame): the target dataframe
    
    Returns
    -----
    None
    
    Example
    -----
    df = pd.DataFrame(data = {"height":[147,190],"weight":[47,72],"age":[12,28]},index = [0,1])
    print_num_col(df) => 
        |
        | "The dataframe has 3 columns now."
        |
    
    
    """
    print(f"The dataframe has {df.shape[1]} columns now.")
    return
        
def print_dim(df: pd.DataFrame) -> None:
    """
    Description
    -----
    Retrieve the shape of dataframe and print it as a statement.
    
    Args
    -----
    df (pd.DataFrame): the target dataframe
    
    Returns
    -----
    None
    
    Example
    -----
    abc_df = pd.DataFrame(data = {"height":[147,190],"weight":[47,72],"age":[12,28]},index = [0,1])
    print_dim(abc_df) =>
        |
        | "There are 2 rows and 3 columns in this dataframe now."
        |
    
    
    """
    print(f"There are {df.shape[0]} rows and {df.shape[1]} columns in this dataframe now.")
    return


def print_null_count(df: pd.DataFrame,cols:list = []) -> None:
    """
    Description
    -----
    Count the null value in each columns.
    If `cols` is provided, only show the null value count for the columns in `cols`. 
    Otherwise, show null value count for all columns.
    
    Args
    -----
    df (pd.DataFrame): target dataframe
    cols (list): the column names to show the null value count. Default: []
    
    Returns
    -----
    None
    
    Example
    -----
    abc_df = pd.DataFrame(data = {"height":[147,190],"weight":[47,np.nan],"age":[np.nan,28]},index = [0,1])
    print_null_count(abc_df) => 
        |
        | === Null Count ===
        | height    0
        | weight    1
        | age       2
        | dtype: int64
        |
        
    print_null_count(abc_df,cols=["age"]) => 
        |
        | === Null Count ===
        | Column `age`: 2
        |
        
    print_null_count(abc_df,cols=["age","weight"]) => 
        |
        | === Null Count ===
        | Column `age`: 2
        | Column `weight`: 1
        |
    
    """
    if len(cols) == 0:
        null_count = df.isnull().sum()
        
        print("=== Null Count ===")
        print(null_count)
    else:
        assert set(cols).issubset(df.columns)
        null_count = df[cols].isnull().sum()
        
        print("=== Null Count ===")
        for col in cols:
            print(f"Column `{col}`: {null_count[col]}")
    return 


def print_null_pct(df: pd.DataFrame,cols:list = []) -> None:
    """
    Description
    -----
    Count the null percentage in each columns.
    If `cols` is provided, only show the null percentage for the columns in `cols`. 
    Otherwise, show null percentage for all columns.
    
    Args
    -----
    df (pd.DataFrame): target dataframe
    cols (list): the column names to show the null percentage. Default: []
    
    Returns
    -----
    None
    
    Example
    -----
    abc_df = pd.DataFrame(data = {"height":[147,190],"weight":[47,np.nan],"age":[np.nan,np.nan]},index = [0,1])
    print_null_pct(abc_df) => 
        |
        | === Null Count Precentage ===
        | height      0.0%
        | weight     50.0%
        | age       100.0%
        | dtype: object
        |
        
    print_null_pct(abc_df,cols=["weight"]) => 
        |
        | === Null Count Precentage ===
        | Column weight: 50.0%
        |
    
    """
    total_row = df.shape[0]
    if len(cols) == 0:
        null_count = df.isnull().sum()
        null_pct = null_count / total_row * 100
        
        print("=== Null Count Precentage ===")
        print(null_pct.round(2).astype(str)+"%")
    else:
        assert set(cols).issubset(df.columns)
        null_count = df[cols].isnull().sum()
        null_pct = null_count / total_row * 100

        print("=== Null Count Precentage ===")
        for col in cols:
            print(f"Column {col}: {round(null_pct[col],4)}%")
    return

def print_duplicated_count(df: pd.DataFrame) -> None:
    """
    Description
    -----
    Count the number of duplicated rows.
    
    Args
    -----
    df (pd.DataFrame): target dataframe
    
    Returns
    -----
    None
    
    Example
    -----
    abc_df = pd.DataFrame(data = {"height":[147,190,147],"weight":[47,np.nan,47],"age":[13,27,13]},index = [0,1,2])
    print_duplicated_count(abc_df) => 
        |
        | There are 1 duplicated rows
        | 
        
    """
    print(f"There are {df.duplicated().sum()} duplicated rows")
    return



### Helper Funcntions (Edit Dataframe)

In [57]:
def drop_cols_if_exist(df: pd.DataFrame,cols_to_drop:list) -> pd.DataFrame:
    """
    Description
    -----
    Drop a column from a dataframe with inplace = True. Only execute the dropping if the cols exist.
    
    Args
    -----
    df (pd.DataFrame): the target dataframe
    cols_to_drop (list): the list of column to be dropped
    
    Returns
    -----
    df (pd.DataFrame): the dataframe with columns dropped
    
    Example
    -----
    # Create a DataFrame
    abc_df = pd.DataFrame(data = {"height":[147,190,147],"weight":[47,np.nan,47],"age":[13,27,13]},index = [0,1,2])
    print(abc_df)  =>
        |
        |    height  weight  age
        | 0     147    47.0   13
        | 1     190     NaN   27
        | 2     147    47.0   13
        |
        
    # Drop columns if exist
    dropped_abc_df = drop_cols_if_exist(abc_df,cols_to_drop=["weight"])
    print(dropped_abc_df)   =>
        | Successfully dropped columns: {'weight'}
        |    height  age
        | 0     147   13
        | 1     190   27
        | 2     147   13
    
    
    """
    intersect_cols = set(cols_to_drop).intersection(df.columns)
    df.drop(columns=intersect_cols,inplace=True)
    print(f"Successfully dropped columns: {intersect_cols}")
    return df    

### General Helper Funcntions

## Data Loading <a class="anchor" id="3-load"></a>

### Load the processed data

In [58]:
# Define dtype for cleaned dataset
PROCESSED_COL_DTYPE = {
    "vin":str,
    "miles":float,
    "year":int,
    "make":str,
    "model":str,
    "trim":str,
    "body_type":str,
    "vehicle_type":str,
    "drivetrain":str,
    "transmission":str,
    "engine_size":float,
    "engine_block":str,
    "city":str,
    "state":str,
    "zip":str,
    "price_range":int,
    
    # For encoded fuel type
    "fuel_M85":int,
    "fuel_Lpg":int,
    "fuel_Diesel":int,
    "fuel_Unleaded":int,
    "fuel_Hydrogen":int,
    "fuel_PremiumUnleaded":int,
    "fuel_Biodiesel":int,
    "fuel_E85":int,
    "fuel_Electric":int,
    "fuel_CompressedNaturalGas":int,
}

In [59]:
# First, we read the cleaned dataset
df = pd.read_parquet(path = "data/processed-data.parquet",
                     columns = PROCESSED_COL_DTYPE)
df.reset_index(drop = True, inplace=True)

### Saity Check

In [60]:
df.head()

vin    miles    year   make model  trim body_type  \
0  19UNC1B01HY800062   9966.0  2017.0  Acura   NSX  Base     Coupe   
1  19UNC1B02HY800023   5988.0  2017.0  Acura   NSX  Base     Coupe   
2  19UNC1B02HY800071  24242.0  2017.0  Acura   NSX  Base     Coupe   
3  19UNC1B02LY800001   6637.0  2020.0  Acura   NSX  Base     Coupe   
4  19UNC1B02LY800001   6637.0  2020.0  Acura   NSX  Base     Coupe   

  vehicle_type drivetrain transmission  engine_size engine_block  \
0          Car        4WD    Automatic          3.5            V   
1          Car        4WD    Automatic          3.5            V   
2          Car        4WD    Automatic          3.5            V   
3          Car        4WD    Automatic          3.5            V   
4          Car        4WD    Automatic          3.5            V   

                  city state      zip  price_range  fuel_M85  fuel_Lpg  \
0           Edmundston    NB  E3V 2K7            5         0         0   
1  Notre-dame-des-pins    QC  G0M 1K0            5         0         0   
2            Coquitlam    BC  V3K 6S4            5         0         0   
3            Pickering    ON  L1V 3P2            5         0         0   
4            Pickering    ON  L1V 3N7            5         0         0   

   fuel_Diesel  fuel_Unleaded  fuel_Hydrogen  fuel_PremiumUnleaded  \
0            0              0              0                     1   
1            0              0              0                     1   
2            0              0              0                     1   
3            0              0              0                     1   
4            0              0              0                     1   

   fuel_Biodiesel  fuel_E85  fuel_Electric  fuel_CompressedNaturalGas  
0               0         0              1                          0  
1               0         0              1                          0  
2               0         0              1                          0  
3               0         0              1                          0  
4               0         0              1                          0

In [61]:
print_dim(df)

There are 4466896 rows and 26 columns in this dataframe now.


In [62]:
print_null_count(df)

=== Null Count ===
vin                          0
miles                        0
year                         0
make                         0
model                        0
trim                         0
body_type                    0
vehicle_type                 0
drivetrain                   0
transmission                 0
engine_size                  0
engine_block                 0
city                         0
state                        0
zip                          0
price_range                  0
fuel_M85                     0
fuel_Lpg                     0
fuel_Diesel                  0
fuel_Unleaded                0
fuel_Hydrogen                0
fuel_PremiumUnleaded         0
fuel_Biodiesel               0
fuel_E85                     0
fuel_Electric                0
fuel_CompressedNaturalGas    0
dtype: int64


In [63]:
print_duplicated_count(df)

There are 0 duplicated rows


## Features Transformation <a class="anchor" id="3-feat-transform"></a>

### Log Transformation on `miles`

In [68]:
df.head()

vin    year   make model  trim body_type vehicle_type  \
0  19UNC1B01HY800062  2017.0  Acura   NSX  Base     Coupe          Car   
1  19UNC1B02HY800023  2017.0  Acura   NSX  Base     Coupe          Car   
2  19UNC1B02HY800071  2017.0  Acura   NSX  Base     Coupe          Car   
3  19UNC1B02LY800001  2020.0  Acura   NSX  Base     Coupe          Car   
4  19UNC1B02LY800001  2020.0  Acura   NSX  Base     Coupe          Car   

  drivetrain transmission  engine_size engine_block                 city  \
0        4WD    Automatic          3.5            V           Edmundston   
1        4WD    Automatic          3.5            V  Notre-dame-des-pins   
2        4WD    Automatic          3.5            V            Coquitlam   
3        4WD    Automatic          3.5            V            Pickering   
4        4WD    Automatic          3.5            V            Pickering   

  state      zip  price_range  fuel_M85  fuel_Lpg  fuel_Diesel  fuel_Unleaded  \
0    NB  E3V 2K7            5         0         0            0              0   
1    QC  G0M 1K0            5         0         0            0              0   
2    BC  V3K 6S4            5         0         0            0              0   
3    ON  L1V 3P2            5         0         0            0              0   
4    ON  L1V 3N7            5         0         0            0              0   

   fuel_Hydrogen  fuel_PremiumUnleaded  fuel_Biodiesel  fuel_E85  \
0              0                     1               0         0   
1              0                     1               0         0   
2              0                     1               0         0   
3              0                     1               0         0   
4              0                     1               0         0   

   fuel_Electric  fuel_CompressedNaturalGas  log_miles  
0              1                          0   9.207035  
1              1                          0   8.697680  
2              1                          0  10.095883  
3              1                          0   8.800566  
4              1                          0   8.800566

In [65]:
# Add 1 to the column to avoid log(0) = -inf
df["log_miles"] = np.log(df["miles"]+1)

In [66]:
# Drop the miles column
drop_cols_if_exist(df,["miles"])

Successfully dropped columns: {'miles'}


vin    year   make             model        trim  \
0        19UNC1B01HY800062  2017.0  Acura               NSX        Base   
1        19UNC1B02HY800023  2017.0  Acura               NSX        Base   
2        19UNC1B02HY800071  2017.0  Acura               NSX        Base   
3        19UNC1B02LY800001  2020.0  Acura               NSX        Base   
4        19UNC1B02LY800001  2020.0  Acura               NSX        Base   
...                    ...     ...    ...               ...         ...   
4466891  1FT7W2BT9KEE35377  2019.0   Ford  F-250 Super Duty         XLT   
4466892  1FD7W2BT0KEC89883  2019.0   Ford  F-250 Super Duty          XL   
4466893  1FT7W2BT5EEB84944  2014.0   Ford  F-250 Super Duty      Lariat   
4466894  1FT7W2BT4KEF92394  2019.0   Ford  F-250 Super Duty      Lariat   
4466895  1FT7W2BT0BEB86340  2011.0   Ford  F-250 Super Duty  King Ranch   

        body_type vehicle_type drivetrain transmission  engine_size  \
0           Coupe          Car        4WD    Automatic          3.5   
1           Coupe          Car        4WD    Automatic          3.5   
2           Coupe          Car        4WD    Automatic          3.5   
3           Coupe          Car        4WD    Automatic          3.5   
4           Coupe          Car        4WD    Automatic          3.5   
...           ...          ...        ...          ...          ...   
4466891    Pickup        Truck        4WD    Automatic          6.7   
4466892    Pickup        Truck        4WD    Automatic          6.7   
4466893    Pickup        Truck        4WD    Automatic          6.7   
4466894    Pickup        Truck        4WD    Automatic          6.7   
4466895    Pickup        Truck        4WD    Automatic          6.7   

        engine_block                 city state      zip  price_range  \
0                  V           Edmundston    NB  E3V 2K7            5   
1                  V  Notre-dame-des-pins    QC  G0M 1K0            5   
2                  V            Coquitlam    BC  V3K 6S4            5   
3                  V            Pickering    ON  L1V 3P2            5   
4                  V            Pickering    ON  L1V 3N7            5   
...              ...                  ...   ...      ...          ...   
4466891            V               Irving    TX    75063            4   
4466892            V              Ephrata    WA    98823            4   
4466893            V                 Pana    IL    62557            4   
4466894            V         Fayetteville    AR    72703            4   
4466895            V                Tampa    FL    33607            3   

         fuel_M85  fuel_Lpg  fuel_Diesel  fuel_Unleaded  fuel_Hydrogen  \
0               0         0            0              0              0   
1               0         0            0              0              0   
2               0         0            0              0              0   
3               0         0            0              0              0   
4               0         0            0              0              0   
...           ...       ...          ...            ...            ...   
4466891         0         0            1              0              0   
4466892         0         0            1              0              0   
4466893         0         0            1              0              0   
4466894         0         0            1              0              0   
4466895         0         0            1              0              0   

         fuel_PremiumUnleaded  fuel_Biodiesel  fuel_E85  fuel_Electric  \
0                           1               0         0              1   
1                           1               0         0              1   
2                           1               0         0              1   
3                           1               0         0              1   
4                           1               0         0              1   
...                       ...             ...       ...    

In [69]:
# Sanity Check
df.head()

vin    year   make model  trim body_type vehicle_type  \
0  19UNC1B01HY800062  2017.0  Acura   NSX  Base     Coupe          Car   
1  19UNC1B02HY800023  2017.0  Acura   NSX  Base     Coupe          Car   
2  19UNC1B02HY800071  2017.0  Acura   NSX  Base     Coupe          Car   
3  19UNC1B02LY800001  2020.0  Acura   NSX  Base     Coupe          Car   
4  19UNC1B02LY800001  2020.0  Acura   NSX  Base     Coupe          Car   

  drivetrain transmission  engine_size engine_block                 city  \
0        4WD    Automatic          3.5            V           Edmundston   
1        4WD    Automatic          3.5            V  Notre-dame-des-pins   
2        4WD    Automatic          3.5            V            Coquitlam   
3        4WD    Automatic          3.5            V            Pickering   
4        4WD    Automatic          3.5            V            Pickering   

  state      zip  price_range  fuel_M85  fuel_Lpg  fuel_Diesel  fuel_Unleaded  \
0    NB  E3V 2K7            5         0         0            0              0   
1    QC  G0M 1K0            5         0         0            0              0   
2    BC  V3K 6S4            5         0         0            0              0   
3    ON  L1V 3P2            5         0         0            0              0   
4    ON  L1V 3N7            5         0         0            0              0   

   fuel_Hydrogen  fuel_PremiumUnleaded  fuel_Biodiesel  fuel_E85  \
0              0                     1               0         0   
1              0                     1               0         0   
2              0                     1               0         0   
3              0                     1               0         0   
4              0                     1               0         0   

   fuel_Electric  fuel_CompressedNaturalGas  log_miles  
0              1                          0   9.207035  
1              1                          0   8.697680  
2              1                          0  10.095883  
3              1                          0   8.800566  
4              1                          0   8.800566

### Same Vehicle History Analysis

In [20]:
# TODO: Same Vehicle History

#### Multiple Listing Vehicles Count

Let's check which cars are with multiple listing.

In [116]:
df["vin"].value_counts()[0:20]

1C6RR6TT7LS102367    68
1C6RR6TT3LS100471    62
1GCGSCEN6J1265015    61
2G1WC5E34G1127360    61
5YJSA1H23EFP62202    61
WBY7Z4C51JVC34561    60
KL4MMBS2XLB097887    60
1C6RR7TT8LS117773    60
WBY7Z4C57JVD95870    60
KL4MMDSL4LB107883    59
JN8AE2KP7H9168001    59
1N4AZ1CP2JC309340    59
1G1YY25U975129704    59
5YJSA1E27GF129998    59
3LN6L5LU4KR613914    58
WBY7Z4C55JVD95415    58
1FT7W2BT5KEF52440    58
5YJ3E1EAXJF160721    58
3GNKBFRS3KS683020    57
WDDZF6EB2HA271060    57
Name: vin, dtype: int64

Let's count how many vehicles have multiple listing.

In [71]:
print(f"There are {(df['vin'].value_counts() > 1).sum()} unique vehicles with more than one listing records.")

There are 761699 unique vehicles with more than one listing records.


#### How to deal with Multiple Listing Vehicles

Let's use one of the examples to analyse what we should do on these multiple listing vehicles.  
Our example: `vin` = `1GCGSCEN6J1265015`

In [23]:
df[df["vin"]=="1GCGSCEN6J1265015"].T

1388721            1391331  \
vin                        1GCGSCEN6J1265015  1GCGSCEN6J1265015   
year                                  2018.0             2018.0   
make                               Chevrolet          Chevrolet   
model                               Colorado           Colorado   
trim                                      LT                 LT   
body_type                             Pickup             Pickup   
vehicle_type                           Truck              Truck   
drivetrain                               RWD                RWD   
transmission                       Automatic          Automatic   
engine_size                              3.6                3.6   
engine_block                               V                  V   
city                                Franklin             Monroe   
state                                     TN                 NC   
zip                                    37067              28110   
price_range                                3                  3   
fuel_M85                                   0                  0   
fuel_Lpg                                   0                  0   
fuel_Diesel                                0                  0   
fuel_Unleaded                              1                  1   
fuel_Hydrogen                              0                  0   
fuel_PremiumUnleaded                       0                  0   
fuel_Biodiesel                             0                  0   
fuel_E85                                   0                  0   
fuel_Electric                              0                  0   
fuel_CompressedNaturalGas                  0                  0   
log_miles                          10.799698                0.0   

                                     1394846            1396780  \
vin                        1GCGSCEN6J1265015  1GCGSCEN6J1265015   
year                                  2018.0             2018.0   
make                               Chevrolet          Chevrolet   
model                               Colorado           Colorado   
trim                                      LT                 LT   
body_type                             Pickup             Pickup   
vehicle_type                           Truck              Truck   
drivetrain                               RWD                RWD   
transmission                       Automatic          Automatic   
engine_size                              3.6                3.6   
engine_block                               V                  V   
city                              Wilmington             Duluth   
state                                     NC                 GA   
zip                                    28405              30096   
price_range                                3                  3   
fuel_M85                                   0                  0   
fuel_Lpg                                   0                  0   
fuel_Diesel                                0                  0   
fuel_Unleaded                              1                  1   
fuel_Hydrogen                              0                  0   
fuel_PremiumUnleaded                       0                  0   
fuel_Biodiesel                             0                  0   
fuel_E85                                   0                  0   
fuel_Electric                              0                  0   
fuel_CompressedNaturalGas                  0                  0   
log_miles                                0.0                0.0   

                                     1402023            1402247  \
vin                        1GCGSCEN6J1265015  1GCGSCEN6J1265015   
year                                  2018.0             2018.0   
make                               Chevrolet          Chevrolet   
model                               Colorado           Colorado   
trim                                      LT                 LT   
body_type      

**Insight:**
- For the same car with `vin` of `1GCGSCEN6J1265015`, there are multiple listing records. We do not have any temporal columns to indicate the time of listing, but we can deduce the temporal order of the listing. 
- Even for the listing with same `log_miles`, they have multiple record due to different locations, different supported fuel types.
    - Different **locations** means the same car is listed through multiple dealers, 
    - Different **supported fuel types** means either one of the following cases:
        1. The supported fuel types were wrongly enter (i.e. car not supporting Diesel => `Diesel = `]).
        2. The supported fuel types were omitted (i.e. car supporting Diesel and Unleaded => `Diesel = 1` & `Unleaded = 0`).
- For the same car listing on multiple platforms, it should have the same `miles` travelled on the odometer. This means we may deduce if the listing are at the same time by the `miles` (or `log_miles`) columns.
- For some of the records, the `engine_block` may even change, it may be caused by the auto-part replacement by the car owner.
    - If these records are with different info of the car, but same `vin` number, then we use majority vote to decide which record to keep.
    - For supported fuel types, as it accept multiple values, we will only keep those support fuel types that with > 50% occurence in all the records with same `log_miles`.

In [72]:
df.loc[df["vin"]=="1GCGSCEN6J1265015","log_miles"].unique()

array([10.79969802,  0.        , 10.6025174 ])

**Insight:**  
From the above example, we can see 3 different values of `log_miles`.

Note that when we applied log-transformation, we added 1 to each values to avoid getting `-inf`. After log-transformation, we had 0, which means it is 0 miles before log-transformation (i.e. `log(0+1) = 0`).

We can deduce the temporal order of listing should be:
1. `log_miles`: 0           (Oldest Listing)
2. `log_miles`: 10.6025174
3. `log_miles`: 10.79969802 (Latest listing)

In [147]:
def process_same_car_history(df,vin):
    same_car_records = df[(df["vin"]==vin)]
    if len(same_car_records) == 0:
        print(f"No records found for vin: {vin}.")
        return
    elif len(same_car_records) == 1:
        print(f"Only 1 records found for vin: {vin}.")
        
    log_miles = sorted(same_car_records["log_miles"].unique(),reverse= False)
    log_miles_dict = {}
    # Iterate each group with different log_miles values
    for lm in log_miles:
        temp_df = same_car_records[same_car_records["log_miles"]==lm]
        
        # Check Duplicates
        uni_records = temp_df.drop(columns=["city","state","zip"]).drop_duplicates()
        log_miles_dict[lm] = uni_records
        
        if len(uni_records) != 1:
            # Process the history if there are different values in the vehicle info / supported fuel types
            col_multiple_dict = {}
            col_replace_dict = {}
            total_records = len(uni_records)
            for col in uni_records.columns:
                col_val_count = uni_records[col].value_counts()
                if len(col_val_count) > 1:
                    
                    mode = uni_records[col].mode()
                    col_multiple_dict[col] = col_val_count
                    idmax = col_val_count.idxmax()
                    
                    col_replace_dict[col] = idmax
                    
                    # if there are multiple values in same columns, use the one with maximum occurence to replace it.
                
                    
                    # print(f"\ncol: {col}")
                    # print(f"argmax:{col_val_count.idxmax()}")
                    # print(col_val_count)
                    # print("---")
                    print(col_replace_dict)
                    print("---")
                    print(f"mode: {mode.values}")
        else:
            print("Records are good")
        
        print(f"\n\nlog_miles: {lm} | Unique: {uni_records.shape[0]}")
      
    
        
        
    return same_car_records

In [148]:
# process_same_car_history(df,"2G1WC5E34G1127360") # engine same
# process_same_car_history(df,"WBY7Z4C51JVC34561") # all the same
process_same_car_history(df,"1C6RR6TT7LS102367") # all the same

{'engine_block': 'V'}
---
mode: ['I' 'V']
{'engine_block': 'V', 'fuel_Diesel': 0}
---
mode: [0]
{'engine_block': 'V', 'fuel_Diesel': 0, 'fuel_Unleaded': 1}
---
mode: [1]
{'engine_block': 'V', 'fuel_Diesel': 0, 'fuel_Unleaded': 1, 'fuel_PremiumUnleaded': 0}
---
mode: [0]
{'engine_block': 'V', 'fuel_Diesel': 0, 'fuel_Unleaded': 1, 'fuel_PremiumUnleaded': 0, 'fuel_E85': 0}
---
mode: [0]
{'engine_block': 'V', 'fuel_Diesel': 0, 'fuel_Unleaded': 1, 'fuel_PremiumUnleaded': 0, 'fuel_E85': 0, 'fuel_Electric': 0}
---
mode: [0]


log_miles: 10.360152617342683 | Unique: 12


vin    year make    model trim body_type vehicle_type  \
370042   1C6RR6TT7LS102367  2020.0  RAM  Classic  SLT    Pickup          Car   
370416   1C6RR6TT7LS102367  2020.0  RAM  Classic  SLT    Pickup          Car   
370602   1C6RR6TT7LS102367  2020.0  RAM  Classic  SLT    Pickup          Car   
370606   1C6RR6TT7LS102367  2020.0  RAM  Classic  SLT    Pickup          Car   
370938   1C6RR6TT7LS102367  2020.0  RAM  Classic  SLT    Pickup          Car   
...                    ...     ...  ...      ...  ...       ...          ...   
3715884  1C6RR6TT7LS102367  2020.0  RAM  Classic  SLT    Pickup          Car   
3717202  1C6RR6TT7LS102367  2020.0  RAM  Classic  SLT    Pickup          Car   
3717311  1C6RR6TT7LS102367  2020.0  RAM  Classic  SLT    Pickup          Car   
3717862  1C6RR6TT7LS102367  2020.0  RAM  Classic  SLT    Pickup          Car   
3717961  1C6RR6TT7LS102367  2020.0  RAM  Classic  SLT    Pickup          Car   

        drivetrain transmission  engine_size engine_block          city state  \
370042         RWD    Automatic          2.5            V          Cary    NC   
370416         RWD    Automatic          2.5            V  Fayetteville    NC   
370602         RWD    Automatic          2.5            I     Charlotte    NC   
370606         RWD    Automatic          2.5            I   Kansas City    MO   
370938         RWD    Automatic          2.5            V    Pleasanton    CA   
...            ...          ...          ...          ...           ...   ...   
3715884        RWD    Automatic          2.5            V     Charlotte    NC   
3717202        RWD    Automatic          2.5            I        Hoover    AL   
3717311        RWD    Automatic          2.5            I       Hickory    NC   
3717862        RWD    Automatic          2.5            V          Cary    NC   
3717961        RWD    Automatic          2.5            V        Monroe    NC   

           zip  price_range  fuel_M85  fuel_Lpg  fuel_Diesel  fuel_Unleaded  \
370042   27511            4         0         0            0              1   
370416   28303            4         0         0            0              1   
370602   28212            4         0         0            0              1   
370606   64153            4         0         0            0              0   
370938   94588            4         0         0            0              0   
...        ...          ...       ...       ...          ...            ...   
3715884  28227            4         0         0            0              1   
3717202  35244            4         0         0            1              0   
3717311  28602            4         0         0            0              1   
3717862  27511            4         0         0            0              0   
3717961  28110            4         0         0            0              1   

         fuel_Hydrogen  fuel_PremiumUnleaded  fuel_Biodiesel  fuel_E85  \
370042               0                     0               0         1   
370416               0                     0               0         0   
370602               0                     0               0         0   
370606               0                     1               0         0   
370938               0                     1               0         0   
...                ...                   ...             ...       ...   
3715884              0                     0               0         0   
3717202              0                     0               0         0   
3717311              0                     0               0         0   
3717862              0                     1               0         0   
3717961              0                     0               0         0   

         fuel_Electric  fuel_CompressedNaturalGas  log_miles  
370042               0                          0  10.360153  
370416               0                          0  10.360153  
370602               0                          0  10.360153  

## Categorical Encoding <a class="anchor" id="3-cat-encode"></a>

In [9]:
# TODO: Categorical Encoding

## Feature Scaling <a class="anchor" id="3-feat-scale"></a>

Available Scaling Method:
- Standard Scaler
- Min-Max Scaler
- Robust Scaler

In [ ]:
# TODO: Scaling data

## Dimensionality Reduction <a class="anchor" id="3-dim-reduct"></a>

Available Dimensionality Reduction Method:
- Principal Component Analysis (PCA)
- Multi-Dimension Scaling (MDS)
- t-Distributed Stochastic Neighbor Embedding (tSNE)

## Feature Selection <a class="anchor" id="3-feat-select"></a>

Available Feature Selection Method:  
[TODO:] Research

## Learning / Takeaway <a class="anchor" id="3-learn"></a>

[Back-to-top](#3-toc)

---

## Reference

#### Columns require encoding
- `make`
- `model`
- `trim`
- `bodt_type`
- `vehicle_type`
- `drivetrain`
- `transmission`
- `engine_block`
- `city`
- `state`
- `zip`

#### Categorical Encoding Methods
After some researches, we concluded that these are the available encoding options for our dataset.  

- `Dummy Encoding` (keep all categories)
- `Dummy Encoding` (with fixed number of category, i.e. top 10 most frequent categories)
- `Dummy Encoding` (with value counts percentage threshold, i.e. only keep categories with more than X% of total records)
- `Ordinal Encoding` (for category with ordinal meaning)
- `Count Encoding`
- `Target Encoding`  

However, some of them are not suitable for our columns.

`Count Enconding`: It is not useful as we have > 7M records and some values may have over millions count.  
`Ordinal Encoding`: Our categorical columns generally do not consist of any order or level, so ordinal enconding may not be useful at all in this scenario.  

For each column, we will explain one by one which enconding methods would be used.

#### Encoding Helper Functions

In [ ]:
def target_enconding(df:pd.DataFrame, feature:str, target:str,replace = True)->pd.DataFrame:
    """ TODO:
    Description
    -----
    
    Args
    -----
    
    Returns
    -----
    
    
    """
    x_len = len(df[feature])
    y_len = len(df[target])
    assert x_len == y_len, "Length of the input_feat should be the same as target_feat"
    
    new_feat_name = f"encode_{feature}"
    count_sum_dict = {}
    for x, y in zip(df[feature], df[target]):
        if x in count_sum_dict:
            # Add the occurence
            count_sum_dict[x][0] += 1
            # Add the value 
            count_sum_dict[x][1] += y
        else:
            count_sum_dict[x] = []
            # Add the occurence with 
            count_sum_dict[x].append(1)
            # Add the value 
            count_sum_dict[x].append(y)
    
    for k,v in count_sum_dict.items():
        # Calculate the mean for each category
        count_sum_dict[k] = count_sum_dict[k][1]/count_sum_dict[k][0]
        
    new_encoded_df = pd.DataFrame([count_sum_dict])\
        .T\
        .reset_index(drop = False)\
        .rename(columns={"index":feature,
                         0:new_feat_name})\
    
    df = df.merge(new_encoded_df,how="left",on=feature)
    
    # Replace the original column if true
    if replace:
        df.drop(columns=[feature],inplace=True)
        df.rename(columns={new_feat_name:feature},inplace=True)
    
    return df

def group_category_by_top_X_summary(df:pd.DataFrame,
                                    col: str,
                                    top_X_list:list = [5,10,15,20,25,30,40,50,100,200,500,1000,2000,5000,10000])-> pd.DataFrame:
    """ TODO:
    Description
    -----
    
    Args
    -----
    
    Returns
    -----
    
    
    """
    assert col in df.columns, f"Column {col} not founded. `col` must be one of the columns in dataframe `df`."
    top_X_df = pd.DataFrame(columns=["top_X",
                                         "standalone_category_sum",
                                         "standalone_category_pct",
                                         "other_sum",
                                         "other_pct",
                                         "total_cat_count"])

    
    
    # Print target column
    print(f"Your target column: {col}")
    
    # Value count for target column
    val_count = df[col].value_counts()
        
    # Test differet top_X
    for top_X in top_X_list:
        if top_X < 0:
            continue
        elif top_X > len(val_count):
            break
            
        this_val_count = val_count[0:top_X]
        other_val_count = val_count[top_X:]

        # Value count percentage for `model` column
        this_val_count_pct = this_val_count/df.shape[0]*100

        standalone_category_sum = this_val_count.sum()
        standalone_category_pct = round(standalone_category_sum/df.shape[0]*100,2)
        other_sum = other_val_count.sum()
        other_pct = round(other_sum/df.shape[0]*100,2)
        total_cat_count = len(this_val_count) + 1
        top_X_df = top_X_df.append({"top_X":top_X,
                                    "standalone_category_sum":standalone_category_sum,
                                    "standalone_category_pct":standalone_category_pct,
                                    "other_sum":other_sum,
                                    "other_pct":other_pct,
                                    "total_cat_count":total_cat_count},ignore_index=True)
    return top_X_df


def group_category_by_threshold_summary(df:pd.DataFrame,col: str,threshold_pct_list:list = [10,5,3,2,1,0.5,0.3,0.2,0.1,0.05,0.03,0.02,0.01])-> pd.DataFrame:
    """ TODO:
    Description
    -----
    
    Args
    -----
    
    Returns
    -----
    
    
    """
    assert col in df.columns, f"Column {col} not founded. `col` must be one of the columns in dataframe `df`."
    assert len(list((filter(lambda x:x < 0,threshold_pct_list)))) + len(list((filter(lambda x:x > 100,threshold_pct_list)))) == 0, "Threshold must be number between 0 and 100"
    assert len(threshold_pct_list) > 0
    threshold_df = pd.DataFrame(columns=["threshold_%",
                                         "standalone_category_sum",
                                         "standalone_category_pct",
                                         "other_sum",
                                         "other_pct",
                                         "total_cat_count"])

    # Value count for target column
    this_val_count = df[col].value_counts()

    # Value count percentage for `model` column
    this_val_count_pct = this_val_count/df.shape[0]*100
    
    # Print target column
    print(f"Your target column: {col}")
    # Test differet threshold
    for threshold in threshold_pct_list:

        standalone_category_sum = this_val_count[this_val_count_pct > threshold].sum()
        standalone_category_pct = round(standalone_category_sum/df.shape[0]*100,2)
        other_sum = this_val_count[this_val_count_pct <= threshold].sum()
        other_pct = round(other_sum/df.shape[0]*100,2)
        total_cat_count = len(this_val_count[this_val_count_pct > threshold]) +1
        threshold_df = threshold_df.append({"threshold_%":threshold,
                                            "standalone_category_sum":standalone_category_sum,
                                            "standalone_category_pct":standalone_category_pct,
                                            "other_sum":other_sum,
                                            "other_pct":other_pct,
                                            "total_cat_count":total_cat_count},ignore_index=True)
    return threshold_df


def group_category_by_top_X(df:pd.DataFrame,col: str,top_X: int,replace = True)-> pd.DataFrame:
    """ TODO:
    Description
    -----
    
    Args
    -----
    
    Returns
    -----
    
    
    """
    assert col in df.columns, f"Column {col} not found. `col` must be one of {df.columns}"
    target_val_count = df[col].value_counts()[:top_X]
    keep_group = list(target_val_count.index)

    df[f"{col}_group"] = df[col].apply(lambda g: g if g in keep_group else "other")
    if replace:
        df.drop(columns=[col],inplace=True)
        df.rename(columns={f"{col}_group":col},inplace = True)
    return df

def group_category_by_pct(df:pd.DataFrame,col: str,threshold_pct: float,replace = True)-> pd.DataFrame:
    """ TODO:
    Description
    -----
    
    Args
    -----
    
    Returns
    -----
    
    
    """
    assert col in df.columns, f"Column {col} not found. `col` must be one of {df.columns}"
    target_val_count = df[col].value_counts() 
    keep_group = list(target_val_count[target_val_count/df.shape[0]*100>threshold_pct].index)

    df[f"{col}_group"] = df[col].apply(lambda g: g if g in keep_group else "other")
    
    if replace:
        df.drop(columns=[col],inplace=True)
        df.rename(columns={f"{col}_group":col},inplace = True)
    return df

from sklearn.preprocessing import LabelBinarizer 
def convert_dummy(df:pd.DataFrame,col:str,concat = True)->pd.DataFrame:
    
    dummy_encoder = LabelBinarizer() 
    
    dummy_output = dummy_encoder.fit_transform(df[col]) 
    dummy_df = pd.DataFrame(dummy_output, 
                            columns = dummy_encoder.classes_)
            
    # Add prefix
    dummy_df = dummy_df.add_prefix(f"{col}_")
    
    # Add the dummy df to original df
    if concat:
        df = pd.concat([df,dummy_df],axis = 1)
    return df

#### Column: `make` 

##### Assessment on column `make`

In [ ]:
make_val_count = encoded_df['make'].value_counts()
make_val_count

In [ ]:
print(f"There are {len(make_val_count)} distinct values in the dataframe.")

##### Summary for dummy encoding by Top X on column `make`

In [ ]:
group_category_by_top_X_summary(encoded_df,"make")

##### Summary for dummy encoding by threshold on column `make`

In [ ]:
group_category_by_threshold_summary(encoded_df,"make")

##### Decision on column `make`
- Both `dummy encoding` methods work fine for the `make` column, but both methods require to group some brand as `Other` group.
- `Dummy encoding` without grouping any categories will generate a huge feature space, which may not be appropriate in the column `make`.
- Vehicle maker / brand (`make`) does a important role when predicting their price, some brands are specialist in luxury cars and the listing record are generally less than other brand.
- While some less popular brands are specialist in cheaper vehicles are also with less listing record. 
- If we apply dummy encoding by a threshold or top X category, we may need to group different brands into the same group, which they may have different market position.
- Therefore, we will go with `target encoding` for this column.

In [ ]:
encoded_df = target_enconding(encoded_df,
                              feature = "make",
                              target = "price_range",
                              replace = True)

In [ ]:
encoded_df.head()

#### Column: `model`

##### Assessment on column `model`

In [ ]:
model_val_count = encoded_df['model'].value_counts()
model_val_count

In [ ]:
print(f"There are {len(model_val_count)} distinct values in the dataframe.")

##### Summary for dummy encoding by Top X on column `model`

In [ ]:
group_category_by_top_X_summary(encoded_df,"model")

##### Summary for dummy encoding by threshold on column `model`

In [ ]:
group_category_by_threshold_summary(encoded_df,"model")

##### Decision on column `model`
- Note that we have so many distinct values in this column.
- If we use `target encoding`, then many values will become approximately zero, which is not something we want. Therefore, we will stick to `dummy encoding`.
- However, it is impossible to keep all categories. The problem is how to determine which categories to keep, by percentage count or by top X.
- As model is also one of the key factor in price listing, we can allow this column to have more dummy column (i.e. 100 columns).
- Let's go with `Top X` approach.

In [ ]:
encoded_df = group_category_by_top_X(encoded_df,col="model",top_X=100,replace = True)

In [ ]:
# Sanity Check
encoded_df.head()

In [ ]:
# Sanity Check
print_dim(encoded_df)

##### Convert the columns into dummy variables

In [ ]:
encoded_df = convert_dummy(encoded_df,"model",concat = True)

In [ ]:
# Sanity Check
print_dim(encoded_df)

##### Avoid Dummy Variable Trap

In [ ]:
cols = list(filter(lambda c: "model_" in c,encoded_df.columns)) 
cols.append("price_range")

encoded_dfoded_df[col].corr()

#### Column: `trim`

##### Assessment on column `trim`

##### Decision on column `trim`

#### Column: `body_type`

##### Assessment on column `body_type`

##### Decision on column `body_type`

#### Column: `vehicle_type`

##### Assessment on column `vehicle_type`

##### Decision on column `vehicle_type`

#### Column: `drivetrain`

##### Assessment on column `drivetrain`

##### Decision on column `drivetrain`

#### Column: `transmission`

##### Assessment on column `transmission`

##### Decision on column `transmission`

#### Column: `engine_block`

##### Assessment on column `engine_block`

##### Decision on column `engine_block`

#### Column: `city`

##### Assessment on column `city`

##### Decision on column `city`

#### Column: `state`

##### Assessment on column `state`

##### Decision on column `state`

#### Column: `zip`

##### Assessment on column `zip`

##### Decision on column `zip`

---

### Same Vehicle History

---

### LAter Code - 2

3 methods:


Columns to handle:
- model (many)
- trim (many)
- body_type (many)
- vehicle_type (2)

- In order to preserve the important information from the columns, we can group the value that with little occurrence into one single group call `Other`.  
- We can set a percentage threshold for the `trim` columns, if that category label has value_count_percentage over the threshold, then we will keep the category, otherwise, we will group it as `Other`.
- While we want to minimize the number of category we generated during the conversion of categorical data into dummy variables, we do not want the percentage of `Other` group to be so high, which may lead to class imbalance.
- This is a tradeoff between **Class Imbanlance** and **Number of dummy variable generated**.

In [ ]:
# Test with differet threshold percentage
# group_category_by_threshold_summary(df,"trim")

- To get an appropriate tradeoff between class imbalance and overwhelming categorical features, we will use `0.3%` as the threshold here.  
- After the dummy variable conversion, we will have 56 standalone categories withdata count percentage > 0.3% and the group `Other` will have ≈ 26.18% of data.

In [ ]:
# df = group_category(df,target_col,0.3)
# df.head()

---

## Neural Network

In [109]:
import pytorch

ModuleNotFoundError: No module named 'pytorch'